common/trinary/flex_trit.c

```diff
@@ -25,65 +25,101 @@ size_t flex_trits_slice(flex_trit_t *const to_flex_trits, size_t const to_len, f
   // num_bytes == num_trits in a 1:1 scheme
   memcpy(to_flex_trits, flex_trits + start, num_bytes);
 #elif defined(FLEX_TRIT_ENCODING_3_TRITS_PER_BYTE)
-  trit_t trits[6] = {0};
-  size_t index = start / 3U;
-  size_t offset = start % 3U;
-  size_t i, j;
-  size_t end_index = NUM_FLEX_TRITS_FOR_TRITS(len) - 1;
-  for (i = index, j = 0; j < num_bytes; i++, j++) {
-    trytes_to_trits(&flex_trits[i], trits, 1);
-    if (offset && i < end_index) {
-      trytes_to_trits(&flex_trits[i + 1], trits + 3, 1);
+  if (0 == (start % NUM_TRITS_PER_FLEX_TRIT)) {
+    // Copy flex trits as bytes
+    memcpy(to_flex_trits, flex_trits + start / NUM_TRITS_PER_FLEX_TRIT, num_trits / NUM_TRITS_PER_FLEX_TRIT);
+    // Handle tail
+    if (0 != num_trits % NUM_TRITS_PER_FLEX_TRIT) {
+      trit_t trits[3] = {0};
+      trytes_to_trits(flex_trits + (start + num_trits) / NUM_TRITS_PER_FLEX_TRIT, trits, 1);
+      trits_to_trytes(trits, to_flex_trits + num_trits / NUM_TRITS_PER_FLEX_TRIT, num_trits % NUM_TRITS_PER_FLEX_TRIT);
+    }
+  } else {
+    trit_t trits[6] = {0};
+    size_t index = start / 3U;
+    size_t offset = start % 3U;
+    size_t i, j;
+    size_t end_index = NUM_FLEX_TRITS_FOR_TRITS(len) - 1;
+    for (i = index, j = 0; j < num_bytes; i++, j++) {
+      trytes_to_trits(&flex_trits[i], trits, 1);
+      if (offset && i < end_index) {
+        trytes_to_trits(&flex_trits[i + 1], trits + 3, 1);
+      }
+      trits_to_trytes(trits + offset, to_flex_trits + j, 3);
+    }
+    // There is a risk of noise after the last trit so we need to clean up
+    uint8_t residual = num_trits % 3U;
+    if (residual) {
+      // size_t tlen = (num_trits - num_bytes * 3 + 3);
+      trits_to_trytes(trits + offset, to_flex_trits + num_bytes - 1, residual);
     }
-    trits_to_trytes(trits + offset, to_flex_trits + j, 3);
-  }
-  // There is a risk of noise after the last trit so we need to clean up
-  uint8_t residual = num_trits % 3U;
-  if (residual) {
-    // size_t tlen = (num_trits - num_bytes * 3 + 3);
-    trits_to_trytes(trits + offset, to_flex_trits + num_bytes - 1, residual);
   }
 #elif defined(FLEX_TRIT_ENCODING_4_TRITS_PER_BYTE)
-  uint8_t buffer = 0;
-  uint8_t tshift = (start & 3) << 1U;
-  uint8_t rshift = (8U - tshift) & 7;
-  size_t index = start >> 2U;
-  size_t i, j;
-  size_t end_index = NUM_FLEX_TRITS_FOR_TRITS(len) - 1;
-  // Calculate the number of bytes to copy over
-  for (i = index, j = 0; j < num_bytes; i++, j++) {
-    buffer = flex_trits[i];
-    buffer = buffer >> tshift;
-    if (rshift && i < end_index) {
-      buffer |= (flex_trits[i + 1] << rshift);
+  if (0 == (start % NUM_TRITS_PER_FLEX_TRIT)) {
+    // Copy flex trits as bytes
+    memcpy(to_flex_trits, flex_trits + start / NUM_TRITS_PER_FLEX_TRIT, num_trits / NUM_TRITS_PER_FLEX_TRIT);
+    // Handle tail
+    if (0 != num_trits % NUM_TRITS_PER_FLEX_TRIT) {
+      uint8_t buffer = 0;
+      uint8_t shift = (4 - num_trits % NUM_TRITS_PER_FLEX_TRIT) << 1U;
```

Comment:
this code is identical to that of residual calculation (line 88)
can you create a static function that handle that?

Reply:
I don't know a nice way to do that. Something like:
```
flex_trit_t flext_trits_te4_zero_high_trits(flex_trit_t f, size_t n) {
  // n <= 4
  size_t shift = (4 - n) << 1U;
  f <<= shift;
  f >>= shift;
  return f;
}
```

Reply:
Actually, I had an idea to refactor/rewrite a bit that code. Maybe try to make it more readable.